### ISSUE TITLE
Using pre-trained word embeddings

### ISSUE THREAD

**Issue by muik** [tensorflow/ranking#142]
I used pre-trained word embeddings to initialize the weights of query and document embeddings like below code.
```
pre_trained_vectors = np.fromfile(path).reshape([-1, dim])
query_embedding_column = tf.feature_column.embedding_column(
    ...,
    initializer=tf.compat.v1.constant_initializer(pre_trained_vectors))
```
As a result, the training converged faster and the metric of result was pretty improved. I think using pre-trained vectors seems to be quite useful.

But the training time has increased.

The reason is that the weights are initialized with every evaluations.
It takes unnecessarily long, because it will be restored from checkpoint file.
And the data of pre-trained vectors is saved to meta file unnecessarily.

I'd like to know the way of initializing weights only at the first time of training.

**Comment by xuanhuiwang** [comment]
I've looked around and this thread in stack overflow may be related: https://stackoverflow.com/questions/48217599/how-to-initialize-embeddings-layer-within-estimator-api/48243086#48243086

**Comment by muik** [comment]
Thank you for the related link. I tried to use `tf.train.Scaffold` for initializing weights like the answer of the link. But scaffold argument could be passed to `tf.estimator.EstimatorSpec`, because the code where `EstimatorSpec` is instantiated is wraped by `_RankingHead` class on [ranking/head.py](https://github.com/tensorflow/ranking/blob/v0.2.0/tensorflow_ranking/python/head.py#L261).

So I had to modify `_RankingHead` class on head.py file to pass the scaffold argument. Now the problem is solved.

In addition, It would be nice to provide a way of passing arguments to creating EstimatorSpec.

**Comment by mulangonando** [comment]
Hi @muik , do you mind sharing with me the code for adding pertained Embeddings and the part of code where you added the scaffolding.  I think I need this but still not able to completely get it up with what you previously shared. I think am missing something.

**Comment by muik** [comment]
Hi, @mulangonando
I'll share my code snippets.

```
from head import create_ranking_head
...

initial_value = np.fromfile(pretrained_vectors_path).reshape([-1, dim])
scope = 'encoding_layer'
name = 'query_tokens_embedding/embedding_weights'

def init_fn(scaffold, sess):
    with tf.compat.v1.variable_scope(scope, reuse=True) as scope:
        weights = tf.compat.v1.get_variable(name, shape=intial_value.shape)
    sess.run(weights.initializer, {weights.initial_value: initial_value})

scaffold = tf.compat.v1.train.Scaffold(init_fn=init_fn)

ranking_head = create_ranking_head(
    ...,
    scaffold=scaffold)

...
```

**head.py** copied and modified from [ranking/head.py](https://github.com/tensorflow/ranking/blob/v0.2.0/tensorflow_ranking/python/head.py#L261)
```
def create_ranking_head(
                        ...,
                        scaffold=None):
  ...
  return _RankingHead(
      ...
      scaffold=scaffold)

class _RankingHead(_AbstractRankingHead):
  def __init__(self,
               ...,
               scaffold=None):
    ...
    self._scaffold = scaffold

  def create_estimator_spec(...):
        ...
        return tf.estimator.EstimatorSpec(
            ...,
            scaffold=self._scaffold)
```

**Comment by mulangonando** [comment]
> Hi, @mulangonando
> I'll share my code snippets.
> 
> ```
> from head import create_ranking_head
> ...
> 
> initial_value = np.fromfile(pretrained_vectors_path).reshape([-1, dim])
> scope = 'encoding_layer'
> name = 'query_tokens_embedding/embedding_weights'
> 
> def init_fn(scaffold, sess):
>     with tf.compat.v1.variable_scope(scope, reuse=True) as scope:
>         weights = tf.compat.v1.get_variable(name, shape=intial_value.shape)
>     sess.run(weights.initializer, {weights.initial_value: initial_value})
> 
> scaffold = tf.compat.v1.train.Scaffold(init_fn=init_fn)
> 
> ranking_head = create_ranking_head(
>     ...,
>     scaffold=scaffold)
> 
> ...
> ```
> 
> **head.py** copied and modified from [ranking/head.py](https://github.com/tensorflow/ranking/blob/v0.2.0/tensorflow_ranking/python/head.py#L261)
> 
> ```
> def create_ranking_head(
>                         ...,
>                         scaffold=None):
>   ...
>   return _RankingHead(
>       ...
>       scaffold=scaffold)
> 
> class _RankingHead(_AbstractRankingHead):
>   def __init__(self,
>                ...,
>                scaffold=None):
>     ...
>     self._scaffold = scaffold
> 
>   def create_estimator_spec(...):
>         ...
>         return tf.estimator.EstimatorSpec(
>             ...,
>             scaffold=self._scaffold)
> ```

Thanks, I'll try this

**Comment by xuanhuiwang** [comment]
@muik, feel free to send a PR to add `scaffold` to `_RankingHead`.

**Comment by xuanhuiwang** [comment]
Close this by now.
